AdvancedFilters: Create workable defaults for Reports that don't have them

psealock opened the pull request:
Fixes https://github.com/woocommerce/woocommerce-admin/issues/6529

Attempts to add advanced filters to reports that don't already have them was causing fatal errors when extension devs used the prescribed methods of extending.

This PR add better defaults and a check to see if any advanced filters have been added. If they have, add a regular filter "Advanced filters" so the newly added filters can be accessed.

### Screenshots

<img width="719" alt="Screen Shot 2021-06-16 at 3 59 01 PM" src="https://user-images.githubusercontent.com/1922453/122155825-ddc7d800-cebb-11eb-834e-32330c947007.png">

### Detailed test instructions:
Instructions follow the first few steps of https://developer.woocommerce.com/2020/02/20/extending-wc-admin-reports in order to replicate Filters alongside Advanced Filters.

1. Create a new plugin if you don't already have one.

```
npm run create-wc-extension
cd ../<my-plugin-name>
npm install
npm start
```
2. Visit the Taxes Report and make sure filters work correctly.
3. Add an advanced filter in JS for the Taxes Report, which does not currently have any.

```js
const addShippedDateAdvancedFilter = ( advancedFilters ) => {
	advancedFilters.filters.Shipped = {
		labels: {
			add: __( 'Shipped Date', 'woocommerce-admin' ),
			remove: __( 'Remove Shipped Date filter', 'woocommerce-admin' ),
			rule: __( 'Select Shipped Date', 'woocommerce-admin' ),
			title: __(
				'{{title}}Shipped Date{{/title}} {{rule /}} {{filter /}}',
				'woocommerce-admin'
			),
			filter: __( 'Select Shipped Date', 'woocommerce-admin' ),
		},
		rules: [
			{
				value: 'between',
				label: __( 'Between', 'woocommerce-admin' ),
			},
		],
		input: {
			component: 'Date',
		},
	};

	return advancedFilters;
};

addFilter(
	'woocommerce_admin_taxes_report_advanced_filters',
	'lab-reporting-extension',
	addShippedDateAdvancedFilter
);
```
4. Visit the Taxes Report again and see the regular filters now have an option "Advanced filters"
5. Add an advanced filter and see the url update with parameters of the advanced filter


<!-- 
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance: to readme.txt under the "unreleased" list at the top of the changelog. If none exists, please add it. Also include PR number.

If changes pertain to a package, also update CHANGELOG.md in the package's folder in a similar manner.--->

jeffstieler reviewed (commented):
This code looks good and tests well, but were the Products and Revenue reports missed?

psealock commented:
Products and Revenue added, this is ready for another look

jeffstieler commented on client/analytics/report/categories/config.js:
```diff
@@ -44,73 +44,86 @@ export const charts = applyFilters( CATEGORY_REPORT_CHARTS_FILTER, [
+		filters: {},
+		title: _x(
+			'Categories Match {{select /}} Filters',
+			'A sentence describing filters for Variations. See screen shot for context: https://cloudup.com/cSsUY9VeCVJ',
```
Variations should be Categories.

jeffstieler commented on client/analytics/report/coupons/config.js:
```diff
@@ -35,68 +35,78 @@ export const charts = applyFilters( COUPON_REPORT_CHARTS_FILTER, [
+		filters: {},
+		title: _x(
+			'Coupons Match {{select /}} Filters',
+			'A sentence describing filters for Variations. See screen shot for context: https://cloudup.com/cSsUY9VeCVJ',
```
Variations should be Coupons.

jeffstieler commented on client/analytics/report/products/config.js:
```diff
@@ -169,12 +169,30 @@ const variationsConfig = {
+		filters: {},
+		title: _x(
+			'Product Match {{select /}} Filters',
+			'A sentence describing filters for Variations. See screen shot for context: https://cloudup.com/cSsUY9VeCVJ',
```
Variations should be Products.

jeffstieler commented on client/analytics/report/products/config.js:
```diff
@@ -169,12 +169,30 @@ const variationsConfig = {
+	{
+		filters: {},
+		title: _x(
+			'Product Match {{select /}} Filters',
```
Should product be pluralized here?

jeffstieler commented on client/analytics/report/revenue/config.js:
```diff
@@ -60,8 +60,32 @@ export const charts = applyFilters( REVENUE_REPORT_CHARTS_FILTER, [
+
+export const filters = applyFilters(
+	REVENUE_REPORT_FILTERS_FILTER,
+	filterValues
```
This isn't a complete config and results in the following console error: `An error occurred while running 'mapSelect': filterConfig.showFilters is not a function`.

jeffstieler commented on client/analytics/report/revenue/config.js:
```diff
@@ -60,8 +60,32 @@ export const charts = applyFilters( REVENUE_REPORT_CHARTS_FILTER, [
+		filters: {},
+		title: _x(
+			'Revenue Match {{select /}} Filters',
+			'A sentence describing filters for Variations. See screen shot for context: https://cloudup.com/cSsUY9VeCVJ',
```
Variations should be Revenue.

jeffstieler commented on client/analytics/report/revenue/config.js:
```diff
@@ -60,8 +60,32 @@ export const charts = applyFilters( REVENUE_REPORT_CHARTS_FILTER, [
+	{
+		filters: {},
+		title: _x(
+			'Revenue Match {{select /}} Filters',
```
Should this be "Revenue Matches"?

jeffstieler commented on client/analytics/report/stock/config.js:
```diff
@@ -10,32 +10,48 @@ const STOCK_REPORT_ADVANCED_FILTERS_FILTER =
+		filters: {},
+		title: _x(
+			'Taxes Match {{select /}} Filters',
+			'A sentence describing filters for Variations. See screen shot for context: https://cloudup.com/cSsUY9VeCVJ',
```
Variations should be Products.

jeffstieler commented on client/analytics/report/stock/config.js:
```diff
@@ -10,32 +10,48 @@ const STOCK_REPORT_ADVANCED_FILTERS_FILTER =
+	{
+		filters: {},
+		title: _x(
+			'Taxes Match {{select /}} Filters',
```
Should this be Products?

jeffstieler commented on client/analytics/report/taxes/config.js:
```diff
@@ -51,49 +51,65 @@ export const charts = applyFilters( TAXES_REPORT_CHARTS_FILTER, [
+		filters: {},
+		title: _x(
+			'Taxes Match {{select /}} Filters',
+			'A sentence describing filters for Variations. See screen shot for context: https://cloudup.com/cSsUY9VeCVJ',
```
Variations should be Taxes.

jeffstieler reviewed (changes requested):
There's a bug with the Revenue report filter config, and some copy-pasta that needs cleaning up. 🍝

psealock commented:
Note to self: This should be a better default

https://github.com/woocommerce/woocommerce-admin/blob/130eb95bae34a8efd83143c52cc08864efc15131/packages/components/src/filters/index.js#L225

psealock commented:
Thanks for the review @jeffstieler.

I fixed the strings, addressed the Revenue Report problem you found and found another error in Stock Report. The filters modify a query param `type`, not `filter`, so if any advancedFilters have been added a new set of filters gets added.

This is ready for another look 🙇🏽‍♂️

jeffstieler reviewed (approved):
Code looks good and this is testing well. 🛳️   it!

psealock commented:
Since this will be cherry-picked, changelog was added as part of https://github.com/woocommerce/woocommerce-admin/pull/7320
